File: demo/golang/custom-golang/device.go
package custom

import (
	dm "github.com/baetyl/baetyl-go/v2/dmcontext"
)

type Device struct {
	cfg  DeviceConfig
	info *dm.DeviceInfo
	cli  *Simulator
}

// NewDevice 此处为模拟设备采集初始化逻辑
func NewDevice(info *dm.DeviceInfo, cfg DeviceConfig) (*Device, error) {
	d := &Device{
		cfg:  cfg,
		info: info,
		cli:  NewSimulator(cfg.Device),
	}
	return d, nil
}

func (d *Device) ReadProperty(name string) (any, error) {
	return d.cli.Get(name)
}

func (d *Device) Set(val float32, idx int) {
	d.cli.Set(val, idx)
}

File: demo/cxx/custom-cxx/src/driver/device.cpp
#include "device.h"

// Constructor
Device::Device(const DeviceConfig &c, DRIVERSDK::DeviceInfo& inf)
        : cfg(c), info(inf), cli(c.getDevice()) {
}

// Member function implementations...
float Device::get(int index) {
    return cli.get(index);
}

void Device::set(int index, float val) {
    cli.set(index, val);
}

const DRIVERSDK::DeviceInfo &Device::getInfo() const {
    return info;
}

void Device::setInfo(const DRIVERSDK::DeviceInfo &inf) {
    this->info = inf;
}
